File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/spots/entryspot/PlayEntrySpot.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkActorCreateResponse;
import systems.zlink.framework.spots.ZLinkEntrySpot;
import systems.zlink.framework.spots.ZLinkEntrySpotContext;
import systems.zlink.samples.tictactoe.server.configuration.PlaySettings;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayActorJoinGameHandler;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayActorObserveMilestoneHandler;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayerWinMilestoneEventHandler;
import systems.zlink.samples.tictactoe.shared.contracts.ObserveMilestoneRes;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerActorCreateReq;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerWinMilestoneEvent;
import systems.zlink.samples.tictactoe.shared.contracts.WinMilestoneNotify;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

// --8<-- [start:doc-entry-spot]
public final class PlayEntrySpot implements ZLinkEntrySpot<PlayActor> {
    private static final Logger LOGGER = LoggerFactory.getLogger(PlayEntrySpot.class);
    private final ZLinkEntrySpotContext context;
    private final PlaySettings settings;
    private final List<PlayActor> milestoneObservers = new ArrayList<>();

    public PlayEntrySpot(ZLinkEntrySpotContext context, PlaySettings settings) {
        this.context = context;
        this.settings = settings;
        // send: JoinGameMsg를 받고 join 완료 뒤 current session으로 결과를 push한다.
        context.handlers().addHandler(PlayActorJoinGameHandler.class);
        // request: ObserveMilestoneReq에 ObserveMilestoneRes로 응답한다.
        context.handlers().addHandler(PlayActorObserveMilestoneHandler.class);
        // subscribe: PlayerWinMilestoneEvent를 받아 observer session에 알린다.
        context.handlers().addHandler(PlayerWinMilestoneEventHandler.class);
    }

    @Override
    public ZLinkEntrySpotContext context() {
        return context;
    }

    @Override
    public CompletionStage<ZLinkActorCreateResponse> onCreateActor(
            PlayActor actor, ZLinkMessage createRequest) {
        if (createRequest.isEmpty()) {
            return CompletableFuture.completedFuture(ZLinkActorCreateResponse.accept());
        }
        PlayerActorCreateReq request = createRequest.decode(PlayerActorCreateReq.class);
        actor.applyPlayer(request.player());
        return CompletableFuture.completedFuture(ZLinkActorCreateResponse.accept());
    }

    // --8<-- [start:doc-ttt-entry-destroy]
    @Override
    public CompletionStage<Void> onJoinedActor(PlayActor actor) {
        if (actor.destroyAfterEntrySpotJoin()) {
            return context.destroyActor(actor)
                    .thenRun(
                            () ->
                                    LOGGER.info(
                                            "tictactoe-lifecycle actor-destroy-complete actor={}",
                                            actor.actorId()));
        }
        return CompletableFuture.completedFuture(null);
    }

    // --8<-- [end:doc-ttt-entry-destroy]

    @Override
    public CompletionStage<Void> onLeaveActor(PlayActor actor) {
        milestoneObservers.removeIf(existing -> existing.actorId().equals(actor.actorId()));
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnectActor(PlayActor actor) {
        actor.markDisconnected();
        milestoneObservers.removeIf(existing -> existing.actorId().equals(actor.actorId()));
        return CompletableFuture.completedFuture(null);
    }

    public ObserveMilestoneRes observeMilestone(PlayActor actor) {
        rememberObserver(actor);
        return new ObserveMilestoneRes(true);
    }

    // --8<-- [start:doc-ttt-milestone-notify]
    public void notifyMilestone(PlayerWinMilestoneEvent event) {
        WinMilestoneNotify payload =
                new WinMilestoneNotify(
                        event.roomId(), event.actorId(), event.displayName(), event.wins());
        for (PlayActor observer : List.copyOf(milestoneObservers)) {
            observer.context().boundSession().send(payload).submit();
        }
    }

    // --8<-- [end:doc-ttt-milestone-notify]

    private void rememberObserver(PlayActor actor) {
        milestoneObservers.removeIf(existing -> existing.actorId().equals(actor.actorId()));
        milestoneObservers.add(actor);
    }
}
// --8<-- [end:doc-entry-spot]

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/spots/entryspot/PlayEntrySpot.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.entryspot

import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpot
import systems.zlink.framework.kotlin.addHandler
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkActorCreateResponse
import systems.zlink.framework.spots.ZLinkEntrySpotContext
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.actors.PlayActor
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayActorJoinGameHandler
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayActorObserveMilestoneHandler
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers.PlayerWinMilestoneEventHandler
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.ObserveMilestoneRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerActorCreateReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerWinMilestoneEvent
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.WinMilestoneNotify

// --8<-- [start:doc-entry-spot]
class PlayEntrySpot(
    override val context: ZLinkEntrySpotContext,
    private val settings: SampleSettings,
) : ZLinkSuspendingEntrySpot<PlayActor>() {
    private val milestoneObservers = mutableListOf<PlayActor>()

    init {
        // send: JoinGameMsg를 받고 join 완료 뒤 current session으로 결과를 push한다.
        context.handlers().addHandler<PlayActorJoinGameHandler>()
        // request: ObserveMilestoneReq에 ObserveMilestoneRes로 응답한다.
        context.handlers().addHandler<PlayActorObserveMilestoneHandler>()
        // subscribe: PlayerWinMilestoneEvent를 받아 observer session에 알린다.
        context.handlers().addHandler<PlayerWinMilestoneEventHandler>()
    }

    override suspend fun onCreateActorSuspending(
        actor: PlayActor,
        createRequest: ZLinkMessage,
    ): ZLinkActorCreateResponse {
        if (createRequest.isEmpty) {
            return ZLinkActorCreateResponse.accept()
        }
        val request = createRequest.decode(PlayerActorCreateReq::class.java)
        actor.applyPlayer(request.player)
        return ZLinkActorCreateResponse.accept()
    }

    // --8<-- [start:doc-ttt-entry-destroy]
    override suspend fun onJoinedActorSuspending(actor: PlayActor) {
        if (actor.destroyAfterEntrySpotJoin) {
            context.destroyActor(actor).await()
            logger.info("tictactoe-lifecycle actor-destroy-complete actor={}", actor.actorId)
        }
    }

    // --8<-- [end:doc-ttt-entry-destroy]

    override suspend fun onLeaveActorSuspending(actor: PlayActor) {
        milestoneObservers.removeIf { it.actorId == actor.actorId }
    }

    override suspend fun onDisconnectActorSuspending(actor: PlayActor) {
        actor.markDisconnected()
        milestoneObservers.removeIf { it.actorId == actor.actorId }
    }

    fun observeMilestone(actor: PlayActor): ObserveMilestoneRes {
        rememberObserver(actor)
        return ObserveMilestoneRes(true)
    }

    // --8<-- [start:doc-ttt-milestone-notify]
    fun notifyMilestone(event: PlayerWinMilestoneEvent) {
        val payload =
            WinMilestoneNotify(
                roomId = event.roomId,
                actorId = event.actorId,
                displayName = event.displayName,
                wins = event.wins,
            )
        milestoneObservers.toList().forEach { observer ->
            observer.context().boundSession().send(payload).submit()
        }
    }

    // --8<-- [end:doc-ttt-milestone-notify]

    private fun rememberObserver(actor: PlayActor) {
        milestoneObservers.removeIf { it.actorId == actor.actorId }
        milestoneObservers += actor
    }

    private companion object {
        val logger = LoggerFactory.getLogger(PlayEntrySpot::class.java)
    }
}
// --8<-- [end:doc-entry-spot]
